### DISCUSSION TITLE
Way to run a match using a keyboard shortcut

### DISCUSSION THREAD

**Discussion by BNolet** [espanso/espanso#1188] [Feature requests and Ideas]
I'm looking to replace AutoHotKey on Windows with espanso but there are a few things I would like to run using a keyboard shortcut rather than having to type. 

Not sure how possible this is.

**Comment by VladimirMikulic** [comment]
@BNolet Espanso is a text expander only, not a whole desktop automation utility :)

**Comment by federico-terzi** [comment]
Hey @BNolet,

Thank you for the feedback! Could you make some example of shortcuts you would like to use?

Cheers :)

**Comment by 4lch4** [comment]
Personally I think tying key-combos together could be rather useful for this. Being able to do something like.. Ctrl+Alt+J to "activate" my JBH package, and then maybe `K` to insert some standard "ok" response message.

So.. `Ctrl + Alt + J` -> `K` inserts the ":jdok" match.

**Comment by BNolet** [comment]
I have a few snippets of text that I insert in various places.

For example, I insert the snippet of text "DOMAIN\username" and the date
beside it very often and I use Ctrl+Alt+Win+I then Ctrl+Alt+Win+D to insert
them into two different excel cells one after the other.

I insert the snippet Brandon Nolet (07/23/20 13:29:30):  (using current
time and date) any time I close an IT ticket into the solution section to
mark the closing time and I use Ctrl+Alt+Win+S to do that.

I hope that clears things up.

On Thu, Jul 23, 2020 at 11:06 AM Devin W. Leaman <notifications@github.com>
wrote:

> Personally I think tying key-combos together could be rather useful for
> this. Being able to do something like.. Ctrl+Alt+J to "activate" my JBH
> package, and then maybe K to insert some standard "ok" response message.
>
> So.. Ctrl + Alt + J -> K inserts the ":jdok" match.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/federico-terzi/espanso/issues/383#issuecomment-663060658>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACNZ3VE2S7YEX7R3DD4OBZTR5BGYJANCNFSM4PE4OWIQ>
> .
>


-- 

— Brandon Nolet

**Comment by BNolet** [comment]
Another example more relevant to Windows, I use Ctrl+F8 on my Apple
keyboard to be able to play/pause tracks on my media player, along with
several others that have symbols on the Apple Keyboard. This I've also got
setup via AutoHotKey. It would be nice to be able to do in espanso to not
need two different solutions. Then again...this isn't necessarily text
expansion at this point.

On Thu, Jul 23, 2020 at 1:31 PM Brandon Nolet <nolet.brandon@gmail.com>
wrote:

> I have a few snippets of text that I insert in various places.
>
> For example, I insert the snippet of text "DOMAIN\username" and the date
> beside it very often and I use Ctrl+Alt+Win+I then Ctrl+Alt+Win+D to insert
> them into two different excel cells one after the other.
>
> I insert the snippet Brandon Nolet (07/23/20 13:29:30):  (using current
> time and date) any time I close an IT ticket into the solution section to
> mark the closing time and I use Ctrl+Alt+Win+S to do that.
>
> I hope that clears things up.
>
> On Thu, Jul 23, 2020 at 11:06 AM Devin W. Leaman <notifications@github.com>
> wrote:
>
>> Personally I think tying key-combos together could be rather useful for
>> this. Being able to do something like.. Ctrl+Alt+J to "activate" my JBH
>> package, and then maybe K to insert some standard "ok" response message.
>>
>> So.. Ctrl + Alt + J -> K inserts the ":jdok" match.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/federico-terzi/espanso/issues/383#issuecomment-663060658>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACNZ3VE2S7YEX7R3DD4OBZTR5BGYJANCNFSM4PE4OWIQ>
>> .
>>
>
>
> --
>
> — Brandon Nolet
>


-- 

— Brandon Nolet

**Comment by federico-terzi** [comment]
Thank you both for the valid feedback. Indeed espanso is not a desktop automation tool, but rather a text expander. I can see how a Shortcut based trigger could be useful, so let's keep track of it on this issue.

> Another example more relevant to Windows, I use Ctrl+F8 on my Apple
keyboard to be able to play/pause tracks on my media player, along with
several others that have symbols on the Apple Keyboard. This I've also got
setup via AutoHotKey. It would be nice to be able to do in espanso to not
need two different solutions. Then again...this isn't necessarily text
expansion at this point.


I definitely see the usefulness of this concept, even though it's unlikely for espanso to become an AutoHotKey alternative on those tasks, as that's outside the scope of it

**Comment by pocc** [comment]
Per this reddit discussion, https://www.reddit.com/r/espanso/comments/qpnk5c/is_it_possible_to_have_the_trigger_be_something/, I landed here.

I think I figured out what I wanted to do:

Using the ASCII values for CTRL characters, this is already possible (in both 0.7.3, v2.0.5, Macos):

    - trigger: "\x01\x02"
      replace: Hex ^A^B
    # ^C^D aded with CTRL+V in vim, these are 2 characters, not 4
    - trigger: "^C^D"
      replace: Raw ^C^D

It doesn't look like it's currently possible to capture ALT, META, or SHIFT keys, as ALT+A, for example, doesn't have a hex value associated with it.

**Comment by pocc** [comment]
Funnily enough, chaining fails when replace has hex codes in it. So it seems trigger can accept hex characters and non-ASCII characters, but replace cannot. The 4 hex codes in the first example are for the fire emoji.

```
    - trigger: "trigger emoji"
      replace: "\xf0\x9f\x94\xa5"
    - trigger: "\xf0\x9f\x94\xa5"
      replace: "Literally fire 🔥"
```

Ditto with European characters:
```
  - trigger: "ashthorn"
    replace: "æÞ"
  - trigger: "æÞ"
    replace: "Literally fire 🔥"
```

**Reply by smeech** [reply]
`replace:` will generate ctrl-characters, but as text, so they tend to appear as dummy characters. They don't invoke the usual program shortcuts. One needs to use something like `xdotool` to simulate key-presses.

**Comment by federico-terzi** [comment]
@pocc Thank you for the feedback! With the current architecture, that requirement would be difficult to implement, but things might change in the future. I'll make sure to let you know if something changes :)

**Comment by pocc** [comment]
I'm curious what the current architecture is. It looks like there are multiple rust keyboard libraries: https://stackoverflow.com/questions/60130532/detect-keydown. Which one does espanso use?

**Comment by federico-terzi** [comment]
> It looks like there are multiple rust keyboard libraries: https://stackoverflow.com/questions/60130532/detect-keydown. Which one does espanso use?

Espanso doesn't use any library for that, it uses a custom implementation :) The detection logic is located in the `espanso-detect` module: https://github.com/federico-terzi/espanso/tree/master/espanso-detect

**Comment by iansedano** [comment]
I have recently moved my text expanding from Autohotkey to Espanso and I am very happy with it. Not to mention now I can have the same config across different OSs 🙌

One thing I still use Autohotkey for is some application launching. I have some shortcuts and all they do is launch an application. I have Python scripts that can launch applications (and do desktop automation if needed), and I can actually launch these scripts with Espanso as-is:

```
  - trigger: ":launchapp"
    replace: "{{script}}"
    vars:
      - name: "script"
        type: script
        params:
          args:
            - python
            - "%CONFIG%/scripts/launch_app.py"
```

The script just doesn't send anything to `stdout` so the trigger is just deleted, but the script does run (only tested on Windows)

I also don't have to focus a text window to launch this (only tested on Windows)

That said, it would be nice to be able to include Ctrl, Alt, Shift, and key combinations in general to do this same thing. That way I could have all my launching shortcuts across OSs (I'd leave the logic to choose the executable inside the Python scripts). Maybe it would end up looking something like this:

```
  - trigger: "<CTRL+ALT+G>"
    replace: "{{script}}"
    vars:
      - name: "script"
        type: script
        params:
          args:
            - python
            - "%CONFIG%/scripts/launch_app.py"
```

**Reply by iansedano** [reply]
Alternatively there is another suggestion for running scripts that makes this look less hacky - https://github.com/espanso/espanso/discussions/1449

**Reply by smeech** [reply]
It is possible to use \<ctrl-keys> as triggers, but with limitations. See https://espanso.org/docs/matches/basics/#keyboard-triggers.
The Shell extension can be even more useful for starting programs, and it isn't necessary to have external scripts, as they can be included in the YAML. The following Python example is awaiting PR into the docs:
```yml
  - trigger: ":pyscript"
    replace: "{{output}}"
    vars:
      - name: output
        type: script
        params:
          args:
            - python
            - -c
            - |
              fruits = ["apple", "banana", "cherry"]
              for x in fruits:
                print(x)
```
The Shell extension can do the same.

**Comment by smeech** [comment]
#1933

**Comment by AucaCoyan** [comment]
Following the discussion in #1933, thanks smeech!
